felix/wireguard/bootstrap.go

```diff
@@ -17,110 +17,401 @@ package wireguard
+	}
+	if peersToValidate == nil {
+		// Peers to validate will be nil if there are no wireguard peers in the kernel that we need to check.
+		log.Debug("Wireguard is not configured locally")
+		return typhas
+	}
+
+	log.Debugf("Filtering typha endpoints for wireguard: %v", typhas)
+
+	var filtered []discovery.Typha
+
+	for _, typha := range typhas {
+		logCxt := log.WithField("typhaAddr", typha.Addr)
+		if typha.NodeName == nil {
+			logCxt.Info("Typha endpoint has no node information - include typha endpoint")
+			filtered = append(filtered, typha)
+			continue
+		}
+
+		typhaNodeName := *typha.NodeName
+		logCxt = logCxt.WithField("typhaNodeName", typhaNodeName)
+		if typhaNodeName == configParams.FelixHostname {
+			// This is a local typha. We should always be able to connect.
+			logCxt.Info("Typha endpoint is local - include typha endpoint")
+			filtered = append(filtered, typha)
+			continue
+		}
+
+		// Get the public key configured for the typha node.
+		typhaNodeKey, err := getPublicKeyForNode(logCxt, typhaNodeName, v3Client)
+		if err != nil {
+			// If we were unable to determine the public key then just include the endpoint.
+			logCxt.WithError(err).Info("Unable to determine public key for node")
+			filtered = append(filtered, typha)
+			continue
+		}
+		logCxt = logCxt.WithField("typhaNodeKey", typhaNodeKey)
+
+		if typhaNodeKey == "" {
+			// There is no key configured, so perhaps this node has wireguard disabled. Include this typha.
+			logCxt.Info("Typha node does not have a wireguard key - include typha endpoint")
+			filtered = append(filtered, typha)
+		} else if peersToValidate.Contains(typhaNodeKey) {
+			// The public key on the typha node is configured in the local routing table. Include this typha.
+			logCxt.Info("Typha node has a wireguard key that is in the local wireguard routing table - include typha endpoint")
+			filtered = append(filtered, typha)
+		} else {
+			// The public key on the typha node is not configured in the local routing table. There is no point in
+			// including this typha because routing will not work and we'll take longer to find a working typha.
+			logCxt.Warning("Typha node has wireguard key that is not in the local wireguard routing table - exclude typha endpoint")
+		}
+	}
+
+	log.Infof("Filtered typha endpoints: %v", filtered)
 
-	if !configParams.WireguardHostEncryptionEnabled && !dbgBootstrap {
+	return filtered
+}
+
+// RemoveWireguardForHostEncryptionBootstrapping removes all wireguard configuration. This includes:
+// - The wireguard public key
+// - The wireguard device (which in turn will delete all wireguard routing rules).
+func RemoveWireguardForHostEncryptionBootstrapping(
+	configParams *config.Config,
+	getNetlinkHandle func() (netlinkshim.Interface, error),
+	calicoClient clientv3.Interface,
+) error {
+	if !configParams.WireguardHostEncryptionEnabled {
+		// HostEncryption is currently enabled in environments by operator rather than through FelixConfiguration.
```

Comment:
same question about this comment - why does felix care about the operator?

Reply:
i'll remove the comment from here.  Perhaps it'll require some docs changes to discuss how to enable and disable host encryption.